mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/service/impl/MobileAcademyServiceImpl.java

```diff
@@ -457,4 +470,60 @@ public String getScoresForUser(Long callingNumber) {
 
         return scores;
     }
-}
+
+    @Override
+    public void updateMsisdn(Long oldCallingNumber, Long newCallingNumber) {
+
+        if (oldCallingNumber.equals(newCallingNumber)) {
+            return;
+        }
+        // Update Msisdn  In MTRAINING_MODULE_BOOKMARK
+        LOGGER.debug("Fetching Bookmarks for Msisdn {}.", oldCallingNumber);
+        List<Bookmark> existingBookmarks = bookmarkService.getAllBookmarksForUser(oldCallingNumber.toString());
+        if (existingBookmarks.size() > 0) {
+            int i;
+            Bookmark bookmark;
+            for (i = 0; i < existingBookmarks.size(); i++) {
+                bookmark = existingBookmarks.get(i);
+                bookmark.setExternalId(newCallingNumber.toString());
+                bookmarkService.updateBookmark(bookmark);
+            }
+            LOGGER.debug("Updated MSISDN {} to {} in {} Bookmarks", oldCallingNumber, newCallingNumber, i);
+        } else {
+            LOGGER.debug("No Bookmarks exists with given Msisdn");
+        }
+
+        // Update Msisdn  In nms_ma_completion_records
+        LOGGER.debug("Fetching Completion records for Msisdn {}.", oldCallingNumber);
+        CompletionRecord completionRecord = completionRecordDataService.findRecordByCallingNumber(oldCallingNumber);
+        if (null == completionRecord) {
+            LOGGER.debug("No CompletionRecord exists with given Msisdn");
+        } else {
+            completionRecord.setCallingNumber(newCallingNumber);
+            completionRecordDataService.update(completionRecord);
+            LOGGER.debug("Updated MSISDN {} to {} in Completion record", oldCallingNumber, newCallingNumber);
+        }
+
+        // Update Msisdn  In MTRAINING_MODULE_ACTIVITYRECORD
+        LOGGER.debug("Fetching Activity records for Msisdn {}", oldCallingNumber);
+        List<ActivityRecord> existingRecords = activityDataService.findRecordsForUser(oldCallingNumber.toString());
+        if (existingRecords.size() > 0) {
+            int i;
+            ActivityRecord activityRecord;
+            for (i = 0; i < existingRecords.size(); i++) {
+                activityRecord = existingRecords.get(i);
+                activityRecord.setExternalId(newCallingNumber.toString());
+                activityDataService.update(activityRecord);
+            }
+            mtrainingModuleActivityRecordAuditDataService.create(new MtrainingModuleActivityRecordAudit(oldCallingNumber, newCallingNumber));
```

Comment:
Since we are working only on the MSISDN change related issues, we added to ActivityRecord. Issues related to Completion record will be fixed in the corresponding tickets.